### PR TITLE
Proposal for remote dev procedures

### PR DESCRIPTION
**Pull request by jotaen4tinypilot**
Proposal to resolve https://github.com/tiny-pilot/tinypilot/issues/1073.

Here is a proposal for how we can re-define / refine our development workflow for on-device testing.

## Some upfront considerations

I realized that the two things that I do most often are:

- Bootstrapping (setting up) new SD cards for doing tests in a clean environment. I routinely wipe my SD cards for testing (depending on what I work on up to several times per day), and in my book it would be cool to have better utilities for this.
- Performing full installs of bundles, both on “fresh” devices, and also on devices that already have an installation. Although the installation of bundles takes a few minutes each time, I think this is the cleanest and safest way currently to test changes on device.

I’m regularly using the [`install-bundle` script](https://github.com/tiny-pilot/tinypilot/blob/master/scripts/install-bundle). On the one hand, this is pretty handy, especially the “fetch from URL” bit; on the other hand, it’s a bit inconvenient to use on fresh SD cards, because the script obviously doesn’t exist on the device. So for my taste, it would work best to facilitate the dev procedures from the dev machine, rather than from the device. That way, we are independent of the device state.

I also rarely build bundles locally, but I mostly download them from CircleCI nowadays. That’s a bit slower, but I find the local process a bit tedious and error prone (you first have to build the Debian package; then copy that over into the `bundler/bundle/` directory; then kick off the `create-bundle` script; then `rsync` the bundle to the device; then install the bundle).

As mentioned in the ticket, I had cobbled together some local scripts over time, with which I’ve tried to automate these procedures as best as possible. For me personally, it would be cool to make these workflows more mature and standardized. But whether or not that make sense would depend on how it works best for everyone else.

This PR demonstrates what I would have in mind, and should serve as starting point for a discussion. I’d be curious for feedback, and curious to learn about your testing workflows. Maybe we can find a good way that works for everyone. If the workflows and preferences are too wide apart, I think it would also be okay to leave the situation as it is right now (but to still update and improve the documentation in any event).

## Proposal

- Introduce a separate class of dev-scripts for helping with remote testing on device (i.e., “remote dev-scripts”). These scripts are run from the local dev machine, and they rely on a certain SSH configuration being present in order for them to work.
  - One is for bootstrapping new devices. I suppose we all use Voyager devices for testing, so we could either default to that, or offer two variants.
  - One is for “pushing” bundles to a device. This also has to sync up the `install-bundle` script along with the bundle, to conveniently allow installing bundles on a fresh device. It would make sense to me to relocate the downloading logic to this script, and keep the “server-side” one more basic.
- Update the documentation in `CONTRIBUTING.md`. The description is currently still quite rough, though. It would also make sense to me to document some “tips” like how to disable the SSH integrity check, because those can be pretty annoying. As far as I’m concerned, I don’t worry too much about the security of my testing device, since it’s inside a local network, and I only connect it to a “non-sensitive” target machine.
<a data-ca-tag href="https://codeapprove.com/pr/tiny-pilot/tinypilot/1407">Review on CodeApprove</a>

### CODE REVIEW

**Comment by mtlynch** [comment]
This mostly sounds fine to me. Your and @jdeanwallace's opinions have more since you use these workflows more often.

## A la carte?

My high-level feedback is that I'd like the scripts to be more a la carte so that they work independently or together.

A frequent scenario for me is that I have a working TinyPilot Pro build, but I want to test a new bundle. So, currently, my process is:

```bash
ssh pilot@tinypilot
sudo /opt/tinypilot/scripts/install-bundle https://example.com/new-bundle.tgz
```

With the proposed changes, I think the process would be (for someone like me who tries to push bundles only occasionally):

```bash
$ ./dev-scripts/remote/push-bundle https://example.com/new-bundle.tgz
Environment variable $TINYTESTPILOT_SSH_HOSTNAME must be specified.

# Oh, whoops, I guess I have to declare that it's tinypilot.
$ TINYTESTPILOT_SSH_HOSTNAME="tinypilot.local" ./dev-scripts/remote/push-bundle https://example.com/new-bundle.tgz
bash: rsync: command not found

# Oh, I guess I need to install rsync on this machine
$ sudo apt install -y rsync

# Now, let me try again.
$ TINYTESTPILOT_SSH_HOSTNAME="tinypilot.local" ./dev-scripts/remote/push-bundle https://example.com/new-bundle.tgz
root@tinypilot's password: 

# Wait, SSH in as root? Oh, right, I have to bootstrap the TinyPilot device.
$ ./dev-scripts/remote/bootstrap pilot@tinypilot
Environment variable $TINYTESTPILOT_SSH_HOSTNAME must be specified.

# Oh, right.
$ source ./dev-scripts/remote/.env.dev
$ ./dev-scripts/remote/bootstrap pilot@tinypilot

# And let's try that first command again
$ ./dev-scripts/remote/push-bundle https://example.com/new-bundle.tgz

# Worked, hopefully?
```

Could we instead make it so that each of the scripts works individually, but you can make it more convenient if you run them all together?

Like

* `./dev-scripts/remote/bootstrap-raspbian`: Enables development on a bare Raspbian system
* `./dev-scripts/remote/bootstrap-existing-tinypilot`: Enables development on an existing TinyPilot device
* `./dev-scripts/remote/push-bundle`: Pushes a local bundle or URL
* `./scripts/install-bundle`: Installs a local bundle or URL

## Rely more on known-installed tools

My other note is that I didn't have rsync installed on my TinyPilot dev machine. My dev machines are all Debian 11 VMs. It's not a big deal for me to install rsync once, but if there are tools that we know would exist on a TinyPilot or Raspbian system, I'd like to lean on those tools rather than use tools from the dev machine (e.g., use curl or wget on the TinyPilot device instead of the dev system).

We can also do wacky things we wouldn't normally do like just install `curl` or `wget` on the test system if we want it and it's not there, whereas we normally wouldn't just install `curl` in the middle of a normal dev script, but it's okay to do that on a test device for development.

**Comment by jdeanwallace** [comment]
@jotaen4tinypilot 

> I realized that the two things that I do most often are...

Yes, I agree and I do the same.

---

> I’m regularly using the [install-bundle script](https://github.com/tiny-pilot/tinypilot/blob/master/scripts/install-bundle).

Yes, same here.

> it’s a bit inconvenient to use on fresh SD cards, because the script obviously doesn’t exist on the device

I don't find this too bad. I normally run the following:

```bash
scp scripts/install-bundle raspberrypi.local:/tmp/
ssh raspberrypi.local
sudo -i
/tmp/install-bundle https://bundle-url
```

I guess I could also just run this instead:

```bash
ssh raspberrypi.local 'sudo bash -s' <scripts/install-bundle https://bundle-url
```

However, when I'm dealing with private tinypilot-pro bundles, I have to download the bundle first:

```bash
scp scripts/install-bundle ~/Downloads/tinypilot-pro-bundle raspberrypi.local:/tmp/
ssh raspberrypi.local
sudo -i
/tmp/install-bundle /tmp/tinypilot-pro-bundle
```

> it would work best to facilitate the dev procedures from the dev machine, rather than from the device.

Yeah, this would help especially when doing trial and error testing so that I don't waste too much time copying files to the device.

---

> I also rarely build bundles locally

Yes, I've never built a bundle locally. CircleCI bundles are convenient for me.

---

> it would be cool to make these workflows more mature and standardized

I'm skeptical that this is possible. I haven't thought about it too much, but each testing situation seems unique depending on what I'm testing and the state of my device/SD-card. I'm worried that the standardized scripts go unused again (over-time).

That being said, I would really like a document for common setups (in terms of settings and ad-hoc commands) that I can refer to when I forget how to do something. For example:
* Flash SD card with vanilla bullseye (what should my ssh settings be again?)
* Install TinyPilot on Voyager (what should my settings be to use CSI capture chip?)
* How do I test a TinyPilot PR on a device?

---

> Introduce a separate class of dev-scripts for helping with remote testing on device... I suppose we all use Voyager devices for testing

I use a normal Raspberry Pi for testing TinyPilot Community builds and a Voyager for testing TinyPilot Pro builds.

> Update the documentation in CONTRIBUTING.md... to document some “tips” like how to disable the SSH integrity check

Yes please! I always get caught out with this.

**Comment by mtlynch** [comment]
I like stuff being encoded as scripts as much as possible, but I agree

Having stuff in `dev-scripts` is generally "free" in that it doesn't end up on production machines, and we're not that worried about breaking.

Can we set the expectation (and document it) that if you're using some sort of dev script, you should commit it to `dev-scripts`? And if it's useful to other people, great, but you're not responsible for making it work for every scenario.

If others want to extend it for other scenarios (e.g., `.local` vs. not), they can take on the task of adapting it to work for both and hopefully not making it too onerous for your original use case.

Maybe if we get to a point where we have like `./dev-scripts/bootstrap-voyager2a-when-jan-is-on-a-mac`, then we revisit putting too much in `dev-scripts`, but I think if any of us are using dev scripts, we might as well commit them in case they're useful to anyone else.
